In [101]:
from collections import defaultdict
import pandas as pd
import numpy as np

## Help with this section of code

In [85]:
'''
Make an empty dictionary with years as keys
fill dictionary values with dataframes for that year
build hierarchal dataframe using keys


nr_dict = defaultdict()
for x in [2010 + x for x in range(8)] :
    nr_dict[x] = pd.read_csv("~/Projects/NiceRide/Nice_Ride_data/"+str(year)+"/Nice_Ride_"
                             +str(year)+"_station_locations.csv")
#NR = pd.concat(nr_dict)
#NR = pd.DataFrame.from_dict(nr_dict, orient='index')
'''
# Need to find a more pythonic way to load in this data
station_10 = pd.read_csv("~/Projects/NiceRide/Nice_Ride_data/2010/Nice_Ride_2010_station_locations.csv")
station_11 = pd.read_csv("~/Projects/NiceRide/Nice_Ride_data/2011/Nice_Ride_2011_station_locations.csv")
station_12 = pd.read_csv("~/Projects/NiceRide/Nice_Ride_data/2012/Nice_Ride_2012_station_locations.csv")
station_13 = pd.read_csv("~/Projects/NiceRide/Nice_Ride_data/2013/Nice_Ride_2013_station_locations.csv")
station_14 = pd.read_csv("~/Projects/NiceRide/Nice_Ride_data/2014/Nice_Ride_2014_station_locations.csv")
station_15 = pd.read_csv("~/Projects/NiceRide/Nice_Ride_data/2015/Nice_Ride_2015_station_locations.csv")
station_16 = pd.read_csv("~/Projects/NiceRide/Nice_Ride_data/2016/Nice_Ride_2016_station_locations.csv")
station_17 = pd.read_csv("~/Projects/NiceRide/Nice_Ride_data/2017/Nice_Ride_2017_station_locations.csv")

## Station 10

In [3]:
station_10.info()
station_10.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65 entries, 0 to 64
Data columns (total 5 columns):
Terminal     65 non-null int64
Station      65 non-null object
Latitude     65 non-null float64
Longitude    65 non-null float64
Nb docks     65 non-null int64
dtypes: float64(2), int64(2), object(1)
memory usage: 2.6+ KB


Terminal                           Station   Latitude  Longitude  Nb docks
0     30000                100 Main Street SE  44.984892 -93.256551        23
1     30001       25th Street & 33rd. Ave. S.  44.957340 -93.223740        15
2     30002                  Augsburg College  44.966220 -93.238400        21
3     30003  Plymouth Ave N. &  N. Oliver Ave  44.991412 -93.306569        23
4     30004            11th Street & Hennepin  44.975340 -93.278690        23

In [4]:
station_10 = station_10.rename(index=str, columns={"Nb docks": "Ndocks",})

## Station 11

In [5]:
station_11.info()
station_11.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 116 entries, 0 to 115
Data columns (total 5 columns):
Terminal        116 non-null int64
Station         116 non-null object
Lat             116 non-null float64
Long            116 non-null float64
Install date    116 non-null object
dtypes: float64(2), int64(1), object(2)
memory usage: 4.6+ KB


Terminal                          Station        Lat       Long  \
0     30000               100 Main Street SE  44.984892 -93.256551   
1     30001        25th Street & 33rd Ave. S  44.957340 -93.223740   
2     30002                 Augsburg College  44.966220 -93.238400   
3     30003  Plymouth Ave. N & Oliver Ave. N  44.991412 -93.306269   
4     30004           11th Street & Hennepin  44.975340 -93.278690   

  Install date  
0     4/8/2011  
1     4/8/2011  
2     4/8/2011  
3     4/9/2011  
4     4/8/2011

In [6]:
station_11 = station_11.rename(index=str, columns={'Lat': 'Latitude',
                                                  'Long': 'Longitude'})

In [7]:
del station_11['Install date']

2011 station data is the only year to explicitly include this data as a feature. Install date will be removed. The approximate install date can be derived from usage statistics if required.

## Station 12

In [25]:
station_12.info()
station_12.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 146 entries, 0 to 145
Data columns (total 5 columns):
Terminal     146 non-null int64
Station      146 non-null object
Latitude     145 non-null float64
Longitude    145 non-null float64
Nb docks     145 non-null float64
dtypes: float64(3), int64(1), object(1)
memory usage: 5.8+ KB


Terminal                           Station   Latitude  Longitude  Nb docks
0     30000                100 Main Street SE  44.984892 -93.256551      27.0
1     30001       25th Street & 33rd. Ave. S.  44.957340 -93.223740      15.0
2     30002                  Augsburg College  44.966220 -93.238400      15.0
3     30003  Plymouth Ave N. &  N. Oliver Ave  44.991412 -93.306269      19.0
4     30004            11th Street & Hennepin  44.975340 -93.278690      23.0

There is a discrepency between the Terminal, Station, Latitude and Longitude, NDocks columns

In [26]:
station_12[station_12['Nb docks'].isna() == True]

Terminal                Station  Latitude  Longitude  Nb docks
82     30082  Not installed in 2012       NaN        NaN       NaN

Terminal 30082 wasn't installed/used in 2012, and may be removed

In [29]:
station_12 = station_12.dropna()
station_12.reset_index(drop=True)
station_12.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 145 entries, 0 to 145
Data columns (total 5 columns):
Terminal     145 non-null int64
Station      145 non-null object
Latitude     145 non-null float64
Longitude    145 non-null float64
Nb docks     145 non-null float64
dtypes: float64(3), int64(1), object(1)
memory usage: 6.8+ KB


In [32]:
station_12 = station_12.rename(index=str, columns={"Nb docks": "Ndocks"})
station_12['Ndocks'] = station_12['Ndocks'].astype(int)
station_12.info()

<class 'pandas.core.frame.DataFrame'>
Index: 145 entries, 0 to 145
Data columns (total 5 columns):
Terminal     145 non-null int64
Station      145 non-null object
Latitude     145 non-null float64
Longitude    145 non-null float64
Ndocks       145 non-null int64
dtypes: float64(2), int64(2), object(1)
memory usage: 6.8+ KB


## Station 11 resolving Ndocks dilemna

Due to station 11 not having a Ndock column, we'll make one using the Ndock values of the previous year (2010), for any terminals that are new for 2011, we'll use the 2012 Ndock data to fill those values.

## Station 13

In [33]:
station_13.info()
station_13.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 170 entries, 0 to 169
Data columns (total 6 columns):
Terminal      170 non-null int64
Station       170 non-null object
Latitude      170 non-null float64
Longitude     170 non-null float64
Nb docks      170 non-null int64
Unnamed: 5    1 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 8.0+ KB


Terminal                           Station   Latitude  Longitude  Nb docks  \
0     30000                100 Main Street SE  44.984892 -93.256551        27   
1     30001       25th Street & 33rd. Ave. S.  44.957340 -93.223740        15   
2     30002                  Augsburg College  44.966220 -93.238400        15   
3     30003  Plymouth Ave N. &  N. Oliver Ave  44.991412 -93.306269        15   
4     30004            11th Street & Hennepin  44.975340 -93.278690        23   

  Unnamed: 5  
0        NaN  
1        NaN  
2        NaN  
3        NaN  
4        NaN

**Before we remove the Unnamed: 5 column lets figure out what non-NaN values it has.**

In [34]:
station_13[station_13['Unnamed: 5'].notna()]

Terminal                                       Station   Latitude  \
167     30167  LaSalle Ave & 9th Street S (10th & Nicollet)  44.975504   

     Longitude  Nb docks                                         Unnamed: 5  
167 -93.275513        19  Station was at 10th & Nicollet (44.97384, -93....

It contains a note that the station was at a different location, this information is already stored in the station name, and in the lat/long difference from previous year. This column is redundant and will be removed.

In [35]:
del station_13['Unnamed: 5']

In [36]:
station_13 = station_13.rename(index=str, columns={"Nb docks": "Ndocks"})

## Station 14

In [38]:
station_14.info()
station_14.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 170 entries, 0 to 169
Data columns (total 5 columns):
Terminal     170 non-null int64
Station      170 non-null object
Latitude     169 non-null float64
Longitude    169 non-null float64
Nb Docks     170 non-null int64
dtypes: float64(2), int64(2), object(1)
memory usage: 6.7+ KB


Terminal                         Station   Latitude  Longitude  Nb Docks
0     30000              100 Main Street SE  44.984892 -93.256551        27
1     30001        25th Street & 33rd Ave S  44.957341 -93.223374        15
2     30002      Riverside Ave & 23rd Ave S  44.967115 -93.240149        15
3     30003  Plymouth Ave N &  N Oliver Ave  44.991412 -93.306269        15
4     30004          11th Street & Hennepin  44.975340 -93.278690        23

There is a discrepancy between Terminal, Station, Nb Docks and Latitude, Longitude

In [43]:
station_14[station_14.Longitude.isna() == True]

Terminal                Station  Latitude  Longitude  Nb Docks
167     30167  Not installed in 2014       NaN        NaN         0

Terminal 30167 wasn't install in 2014, this observation will be removed

In [44]:
station_14 = station_14.dropna()
station_14.reset_index(drop=True)
station_14.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 169 entries, 0 to 169
Data columns (total 5 columns):
Terminal     169 non-null int64
Station      169 non-null object
Latitude     169 non-null float64
Longitude    169 non-null float64
Nb Docks     169 non-null int64
dtypes: float64(2), int64(2), object(1)
memory usage: 7.9+ KB


In [45]:
station_14 = station_14.rename(index=str, columns={"Nb Docks": "Ndocks"})

## Station 15

In [47]:
station_15.info()
station_15.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190 entries, 0 to 189
Data columns (total 6 columns):
Terminal     190 non-null int64
Station      190 non-null object
Latitude     190 non-null float64
Longitude    190 non-null float64
NbDocks      190 non-null int64
Notes        22 non-null object
dtypes: float64(2), int64(2), object(2)
memory usage: 9.0+ KB


Terminal                         Station   Latitude  Longitude  NbDocks  \
0     30000              100 Main Street SE  44.984892 -93.256551       27   
1     30001        25th Street & 33rd Ave S  44.957341 -93.223374       15   
2     30002      Riverside Ave & 23rd Ave S  44.967115 -93.240149       15   
3     30003  Plymouth Ave N &  N Oliver Ave  44.991412 -93.306269       15   
4     30004          11th Street & Hennepin  44.975340 -93.278690       23   

  Notes  
0   NaN  
1   NaN  
2   NaN  
3   NaN  
4   NaN

In [48]:
station_15 = station_15.rename(index=str, columns={"NbDocks": "Ndocks",})

In [50]:
station_15[station_15['Notes'].notna()]

Terminal                              Station   Latitude  Longitude  \
55      30055       Malcolm Ave SE & 5th Street SE  44.973371 -93.211264   
169     30169     Marshall Street NE & 18th Ave NE  45.006229 -93.271571   
170     30170                      Nicollet Island  44.985388 -93.261450   
171     30171       SE 3rd Ave & University Ave SE  44.985465 -93.252102   
172     30172           2nd Street SE & 5th Ave SE  44.983321 -93.249653   
173     30173   22nd Ave NE & California Street NE  45.010119 -93.268598   
174     30174            Fort Snelling LRT Station  44.892483 -93.197820   
175     30175       18th Ave Trail & 2nd Street NE  45.006321 -93.266553   
176     30176  Washington Ave N & Cedar Lake Trail  44.985406 -93.274031   
177     30177         3rd Street NE & Lowry Ave NE  45.013040 -93.264573   
178     30178           5th Street NE & 1st Ave NE  44.990328 -93.254295   
179     30179             1st Ave N & 2nd Street N  44.983902 -93.268604   
180     30180             3rd Street N & 8th Ave N  44.987996 -93.279854   
181     30181             5th Ave N & 4th Street N  44.984460 -93.277234   
182     30182                         Sanford Hall  44.980831 -93.240282   
183     30183                   Weisman Art Museum  44.973607 -93.237254   
184     30184      Hennepin Ave & S Washington Ave  44.981895 -93.269195   
185     30185           5th Street NE & 3rd Ave NE  44.991821 -93.255558   
186     30186           5th Street & Nicollet Mall  44.978665 -93.270503   
187     30187        Groveland Ave & Clifton Place  44.966145 -93.286294   
188     30188               8th Street & 2nd Ave S  44.974795 -93.270649   
189     30189         Delaware St SE & 26th Ave SE  44.972128 -93.221021   

     Ndocks                                              Notes  
55       25  Moved 6/4/15 from 4th and Hennepin  44.980159,...  
169      15                              Approx. Install Dates  
170      15                                          7/22/2015  
171      15                                          7/23/2015  
172      15                                          7/23/2015  
173      15                                           8/6/2015  
174      15                                          8/27/2015  
175      15                                           8/7/2015  
176      19                                          8/19/2015  
177      15                                          8/22/2105  
178      15                                           8/3/2015  
179      19                                           8/5/2015  
180      15                                          7/30/2015  
181      19                                          7/30/2015  
182      17                                           9/1/2015  
183      19                                           9/3/2015  
184      19                                          8/20/2015  
185      15                                           8/3/2015  
186      19                                          7/20/2015  
187      19                                          7/31/2015  
188      19                                          7/19/2015  
189      19                                          7/24/2015

Notes include install dates, usage data can show approximately when the station became in use. There was a station that was moved this year. Due to the fact that we don't have the data on when it was moved, usage data may be able to determine this information. For the purposes that we're using the data for, this isn't necessary information and can be dropped.

In [53]:
del station_15['Notes']
station_15.info()

KeyError: 'Notes'

## Station 16

In [75]:
station_16.info()
station_16.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 202 entries, 0 to 201
Data columns (total 6 columns):
Terminal      202 non-null object
Station       202 non-null object
Latitude      202 non-null float64
Longitude     202 non-null float64
Nb Docks      202 non-null int64
Unnamed: 5    4 non-null object
dtypes: float64(2), int64(1), object(3)
memory usage: 9.5+ KB


Terminal                       Station   Latitude  Longitude  Nb Docks  \
0    30000            100 Main Street SE  44.984892 -93.256551        27   
1    30039   10th Street & Nicollet Mall  44.973839 -93.274544        11   
2    30117  10th Street E & Cedar Street  44.950189 -93.097590        15   
3    30154     11th Ave S & S 2nd Street  44.976730 -93.252189        23   
4    30004        11th Street & Hennepin  44.975340 -93.278690        23   

  Unnamed: 5  
0        NaN  
1        NaN  
2        NaN  
3        NaN  
4        NaN

In [87]:
station_16[station_16['Unnamed: 5'].notna()]

Terminal          Station   Latitude  Longitude  Nb Docks  \
97     30084   Grand & Cretin  44.939936 -93.192758        15   
150    30146      Webber park  45.033069 -93.290650        15   

                                         Unnamed: 5  
97   Moved Aug 19th to Hennepin Ave & N 5th Street   
150  Moved Aug 9th North Mississippi Regional Park

The moved data here does include note on the date it was moved. For simplicity of analysis, we'll remove the original location "xxxxx-A". This will allow us to have the terminal ID's as ints, and remove the Unnamed: 5 column.

In [77]:
station_16 = station_16.rename(index=str, columns={"Nb docks": "Ndocks"})

In [88]:
del station_16['Unnamed: 5']

In [86]:
station_16.drop(station_16.index[149], inplace = True)
station_16.drop(station_16.index[96], inplace = True)

In [102]:
np.sort(station_16.Terminal.values)

array(['30000', '30001', '30002', '30003', '30004', '30005', '30006',
       '30007', '30008', '30009', '30010', '30011', '30012', '30013',
       '30014', '30015', '30016', '30017', '30018', '30019', '30020',
       '30021', '30022', '30023', '30024', '30025', '30026', '30027',
       '30028', '30029', '30030', '30031', '30032', '30033', '30034',
       '30035', '30036', '30037', '30038', '30039', '30040', '30041',
       '30042', '30043', '30044', '30045', '30046', '30047', '30048',
       '30049', '30050', '30051', '30052', '30053', '30054', '30055',
       '30056', '30057', '30058', '30059', '30060', '30061', '30062',
       '30063', '30064', '30065', '30066', '30067', '30068', '30069',
       '30070', '30071', '30072', '30073', '30074', '30075', '30076',
       '30077', '30078', '30079', '30080', '30081', '30082', '30083',
       '30084', '30085', '30086', '30087', '30088', '30089', '30090',
       '30091', '30092', '30093', '30094', '30095', '30096', '30097',
       '30098', '300

Need to determine what to do with 'NRHQ' terminal ID

In [89]:
# station_16.Terminal = station_16.Terminal.astype(int)

ValueError: invalid literal for int() with base 10: 'NRHQ '

## Station 17

In [103]:
station_17.info()
station_17.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 202 entries, 0 to 201
Data columns (total 5 columns):
Number         202 non-null object
Name           202 non-null object
Latitude       202 non-null float64
Longitude      202 non-null float64
Total docks    202 non-null int64
dtypes: float64(2), int64(1), object(2)
memory usage: 8.0+ KB


Number                            Name   Latitude  Longitude  Total docks
0  30000              100 Main Street SE  44.984892 -93.256551           27
1  30001        25th Street & 33rd Ave S  44.957341 -93.223374           15
2  30002      Riverside Ave & 23rd Ave S  44.967115 -93.240149           15
3  30003  Plymouth Ave N &  N Oliver Ave  44.991412 -93.306269           15
4  30004          11th Street & Hennepin  44.975340 -93.278690           23

In [105]:
station_17 = station_17.rename(index=str, columns={'Total docks': 'Ndocks',
                                                  'Name': 'Station',
                                                  'Number': 'Terminal'})

In [107]:
station_17.info()

<class 'pandas.core.frame.DataFrame'>
Index: 202 entries, 0 to 201
Data columns (total 5 columns):
Terminal     202 non-null object
Station      202 non-null object
Latitude     202 non-null float64
Longitude    202 non-null float64
Ndocks       202 non-null int64
dtypes: float64(2), int64(1), object(2)
memory usage: 9.5+ KB


Terminal column is of type object, need to investigate further:

In [ ]:
station_17.